grobid-core/src/main/java/org/grobid/core/jni/JEPThreadPool.java

```diff
@@ -0,0 +1,116 @@
+package org.grobid.core.jni;
+
+import java.util.concurrent.*;  
+import java.util.*;
+import java.io.*;
+
+import org.grobid.core.utilities.GrobidProperties;
+import org.grobid.core.exceptions.GrobidResourceException;
+
+import jep.Jep;
+import jep.JepConfig;
+import jep.JepException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * For using DeLFT deep learning models, we use JEP as JNI CPython interpreter. 
+ * JEP presents the following constraint: A thread that creates a JEP instance 
+ * must be reused for all method calls to that JEP instance. For ensuring this,
+ * we pool the Jep instances in a singleton class. 
+ */
+
+public class JEPThreadPool { 
+    private static final Logger LOGGER = LoggerFactory.getLogger(JEPThreadPool.class);
+
+    private int POOL_SIZE = 1;
+
+    private ExecutorService executor;  
+    private Map<Long,Jep> jepInstances;
+
+    private static volatile JEPThreadPool instance;
+
+    public static JEPThreadPool getInstance() {
+        if (instance == null) {
+            getNewInstance();
+        }
+        return instance;
+    }
+
+    /**
+     * Creates a new instance.
+     */
+    private static synchronized void getNewInstance() {
+        LOGGER.debug("Get new instance of JEPThreadPool");
+        instance = new JEPThreadPool();
+    }
+
+    /**
+     * Hidden constructor
+     */
+    private JEPThreadPool() {
+        // creating a pool of POOL_SIZE threads
+        //executor = Executors.newFixedThreadPool(POOL_SIZE); 
+        executor = Executors.newSingleThreadExecutor();
+        // each of these threads is associated to a JEP instance
+        jepInstances = new HashMap<Long,Jep>();
```

Comment:
As far as I am aware, [HashMap](https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html) is not thread-safe. Not sure if you have synchronisation further up. There is [ConcurrentHashMap](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html) if you really need it (without synchronization).